### PR TITLE
Story2 t3 get data

### PR DESCRIPTION
**Pull request by davin2020**
Peer review for getting data out of DB using ProductHydrator and BaseProduct classes

### CODE REVIEW

File: src/DbConnector.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace RobotStores;
+// FYI need '\PDO' in order to refer to global  namespace of PHP where PDO lives
+
+class DbConnector
+{
+    static private $db;
```
Comment (jennymegan) [inline src/DbConnector.php]: type hint needed for property

File: src/DbConnector.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    static public function getConnection(): \PDO
```
Comment (jennymegan) [inline src/DbConnector.php:10]: do we need a \ in the typehint? no idea!

File: src/DbConnector.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        if ( self::$db == null ) {
+            $db = new \PDO('mysql:host=db;dbname=robot-stores-dav', 'root', 'password');
```
Comment (jennymegan) [inline src/DbConnector.php]: please replace the placeholder db name here

File: src/Entities/BaseProduct.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+namespace RobotStores\Entities;
+
+class BaseProduct
+{
+    protected int $id;
+    protected string $title;
+    protected float $price;
+    protected string $image;
+
+    public function getID():int {
```
Comment (jennymegan) [inline src/Entities/BaseProduct.php]: missing space before return type

File: src/Entities/BaseProduct.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        return $this->id;
+    }
+
+    public function getTitle():  string {
```
Comment (jennymegan) [inline src/Entities/BaseProduct.php]: maybe that space moved down here.. there seem to be two before this return type :D

File: src/Hydrators/ProductHydrator.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace RobotStores\Hydrators;
+
+class ProductHydrator
+{
+    private array $productsArray;
+
+    static public function getFilteredProducts(int $characterId = null):Array {
```
Comment (jennymegan) [inline src/Hydrators/ProductHydrator.php]: perhaps remove the parameter $ character id altogether here, as it's not part of this story?
also space before return type

File: src/Entities/BaseProduct.php (hunk 4 above)
Comment (jennymegan) [reply src/Entities/BaseProduct.php]: also drop the curly boi brace down onto the next line for method definitions, same as for classes

File: src/Entities/BaseProduct.php (hunk 5 above)
Comment (jennymegan) [reply src/Entities/BaseProduct.php]: drop down curly brace here - and throughout the other files too

File: src/Hydrators/ProductHydrator.php [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        $productsArray = self::getAllProducts();
+        return $productsArray;
+    }
+
+    static private function getAllProducts():Array {
```
Comment (jennymegan) [inline src/Hydrators/ProductHydrator.php]: space before return type

File: src/Hydrators/ProductHydrator.php (hunk 6 above)
Comment (jennymegan) [reply src/Hydrators/ProductHydrator.php]: return type is capitalised here and nowhere else

File: src/Hydrators/ProductHydrator.php (hunk 7 above)
Comment (jennymegan) [reply src/Hydrators/ProductHydrator.php]: this one's capitalised too

File: src/Hydrators/ProductHydrator.php [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        $dbConnection = \RobotStores\DbConnector::getConnection();
+        $query = $dbConnection->prepare('SELECT `id`, `title`, `price`, `image` FROM `products`');
```
Comment (jennymegan) [inline src/Hydrators/ProductHydrator.php]: add a semi colon at end of select statement

**Review by jennymegan — CHANGES_REQUESTED** [review]

File: src/DbConnector.php (hunk 2 above)
Comment (elstevens23) [reply src/DbConnector.php:10]: we're not 100% sure, but we think we need it - PHP Storm doesn't seem to be very happy if we take away the back slash. We need it in order to refer to the global namespace.

Comment (ellie-r) [reply src/DbConnector.php:10]: cool happy to leave as is

File: src/DbConnector.php [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+// FYI need '\PDO' in order to refer to global namespace of PHP where PDO lives
+
+class DbConnector
+{
+    static private \PDO $db;
+
+    static public function getConnection(): \PDO
+    {
+        if ( self::$db == null ) {
+            $db = new \PDO('mysql:host=db;dbname=robot-stores', 'root', 'password');
```
Comment (ellie-r) [inline src/DbConnector.php]: db name needs changing

**Review by ellie-r — CHANGES_REQUESTED** [review]
just one change :)

**Review by ellie-r — APPROVED** [review]
amazing work

**Review by jennymegan — APPROVED** [review]

**Pull request merged**